keps/sig-cluster-lifecycle/0032-create-a-k8s-io-component-repo.md

```diff
@@ -0,0 +1,245 @@
+
+#### Standardized authentication / authorization
+
+In order to not expose this kind of information (e.g. metrics) to anyone that can talk to the component, it may utilize SubjectAccessReview requests to the API server, and hence delegate authentication and authorization to the API server. It should be easy to add this functionality to your component.
+
+### Part 4: Sample implementation in k8s.io/sample-component
+
+Provides an example usage of the three main functions of the _k8s.io/component_ repo, implementing ComponentConfig, the CLI wrapper tooling and the common HTTPS endpoints with delegated auth.
+
+### Code structure
+
+- k8s.io/component
+  - config/
+    - Would hold internal, shared ComponentConfig types across core components
+    - {v1,v1beta1,v1alpha1}
+      - Would hold external, shared ComponentConfig types across core components
+    - serializer/
+      - Would hold common methods for encoding/decoding ComponentConfig
+    - testing/
+      - Would hold common testing code for use in unit tests local to the implementation of ComponentConfig.
+  - cli/
+    - Would hold common methods and types for building a k8s component command (building on top of github.com/spf13/{pflag,cobra})
+    - options/
+      - Would hold flag definitions
+    - testing/
+      - Would hold common testing code for use in unit tests local to the implementation of the code
+    - logging/
+      - Would hold common code for using _k8s.io/klog_
+  - server/
+    - auth/
+      - Would hold code for implementing delegated authentication and authorization to Kubernetes
+    - configz/
+      - Would hold code for implementing a `/configz` endpoint in the component
+    - healthz/
+      - Would hold code for implementing a `/healthz` endpoint in the component
+    - metrics/
+      - Would hold code for implementing a `/metrics` endpoint in the component
+    - pprof/
+      - Would hold code for implementing a `/pprof` endpoint in the component
+    - version/
+      - Would hold code for implementing a `/version` endpoint in the component
+
+### Timeframe and Implementation Order
+
+**Objective:** The ComponentConfig part done for v1.14
+
+**Stretch goal:** Get the CLI and HTTPS server parts done for v1.14.
+
+**Implementation order:**
+
+1. Create the k8s.io/component repo with the initial ComponentConfig shared code
+2. Move shared v1alpha1 ComponentConfig types and references from `k8s.io/api{server,machinery}/pkg/apis/config` to `k8s.io/component/config`
+3. Set up good unit testing for all core ComponentConfig usage, by writing the `k8s.io/component/config/testing` package
+4. Move server-related util packages from `k8s.io/kubernetes/pkg/util/` to `k8s.io/component/server`. e.g. delegated authn/authz "/configz", "/healthz", and "/metrics" packages are suitable
+5. Move common flag parsing / cobra.Command setup code to `k8s.io/component/cli` from (mainly) the kubelet codebase.
```

Comment:
And also try to figure out cleaner ways to do it beforehand, if possible.
IIRC @sttts had some alternative patterns to the Kubelet's approach for config merging that looked nice.